### PR TITLE
Basic testing for Election, Audit, and Contest

### PR DESCRIPTION
**Pull request by smorin8674**
- [X] Basic testing for Audit creation
- [X] Basic testing for Contest creation
- [X] Basic testing for Election creation
- [X] Functions to generate contests and elections (for stress testing later)
- [X] Functions to parse contest and election JSON files
- [X] Templates for Contest and Election JSON
- [X] Contest can no longer be None when creating an Audit

### CODE REVIEW

File: src/r2b2/contest.py [hunk 1]
```diff
@@ -7,7 +7,8 @@
 class ContestType(Enum):
     """Enum indicating what type of vote variation was used in the contest.
 
-    TODO: Add additional vote variations from VVSG.
+    Todo:
+        Add additional vote variations from VVSG.
```
[the remark this answers is not in the document]
Comment (smorin8674) [reply src/r2b2/contest.py]: I only changed it because I the `Todo: `  is an accepted keyword for Sphinx with our style of Docstrings, while `TODO: ` is not and simply appears as plain text in the Documentation. Using this format, it will not be generated in the Documentation on RTD, even though it appears in the Docstring. 

We could switch to `# TODO: ` comments below the Docstring entirely.

[the remark this answers is not in the document]
Comment (nealmcb) [reply src/r2b2/contest.py]: Hmm - puzzling. When I look at r2b2/dist/docs/autoapi/r2b2/contest/index.html I see nothing in the rendered documentation relating to Todo:, though I see it via the [source] link. But with TODO: as before I see it as text in the documentation.

[the remark this answers is not in the document]
Comment (smorin8674) [reply src/r2b2/contest.py]: Exactly, the `Todo: ` will be ignored as it is not useful documentation for the user, but a note for developers.

File: src/r2b2/tests/test_audit.py [hunk 2]
```diff
@@ -18,55 +22,64 @@ def compute_sample_size(self):
 
 
 def test_simple_audit():
```
[the remark this answers is not in the document]
Comment (smorin8674) [reply src/r2b2/tests/test_audit.py]: I agree that we will need `__str__` and `__repr__` methods for our classes. 

With respect to doctests, I was under the impression that doctests are not useful for unit testing because they are less efficient and do not maintain the setup between doctests which would make writing tests for object creation rather difficult. While it is possible to use them for unit testing, I usually only see doctests used for documenting examples of code usage, not for unit testing.

File: src/r2b2/tests/test_audit.py [hunk 3]
```diff
@@ -1,14 +1,18 @@
...
 
+import r2b2.tests.util as util
 from r2b2.audit import Audit
 from r2b2.contest import Contest
 
+default_contest = util.generate_contest(100)
+
 
 class SimpleAudit(Audit):
     """For testing purposes only."""
```
[the remark this answers is not in the document]
Comment (smorin8674) [reply src/r2b2/tests/test_audit.py]: The creation of the new class was precisely the point of the test. Because the `Audit` class is an abstract base class, ensuring that a subclass can properly be created from `Audit`, invoke `super().__init__()`, and create unique implementations of the abstract functions seemed like an important test.

File: src/r2b2/contest.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nealmcb) [reply src/r2b2/contest.py]: Ahh - I misunderstood. Thanks. Good to know we have both options.
I'm guessing that this particular Todo is more of a research agenda for now, working with the SHANGRLA framework. Perhaps noting that as an issue would be more helpful.

File: src/r2b2/tests/test_audit.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nealmcb) [reply src/r2b2/tests/test_audit.py]: Ahh - noting that reasoning may be all we need.
But wouldn't we get more meaningful testing coverage of that init code for free when testing the other classes we derive from Audit?

File: src/r2b2/tests/test_audit.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nealmcb) [reply src/r2b2/tests/test_audit.py]: I think this doctest would be a simple two-liner: invocation and expected result, and would replace the whole 10-line function. It would also provide an example which is easy for users to copy-paste and play with.

I agree that in general, complicated testing needs more intricate setups in unit tests.

File: src/r2b2/tests/util.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
+import json
+from random import randint
+
+from r2b2.contest import Contest
+from r2b2.contest import ContestType
+from r2b2.election import Election
+
+
+def generate_contest(max_size):
+    """Generate a Contest with random data. For testing prposes only.
```
Comment (nealmcb) [inline src/r2b2/tests/util.py]: typo: purposes

File: src/r2b2/tests/util.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    Note:
+        Currently this only generates 2 candidate Pluraity contests with 1 winner.
```
Comment (nealmcb) [inline src/r2b2/tests/util.py]: typo: plurality

File: src/r2b2/tests/util.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    """
+    contest_ballots = randint(1, max_size)
+    win_tally = randint(contest_ballots // 2, contest_ballots)
```
Comment (nealmcb) [inline src/r2b2/tests/util.py]: > win_tally = randint(math.ceil(contest_ballots / 2), contest_ballots)

File: src/r2b2/tests/util.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
+import json
+from random import randint
+
+from r2b2.contest import Contest
+from r2b2.contest import ContestType
+from r2b2.election import Election
+
+
+def generate_contest(max_size):
```
Comment (nealmcb) [inline src/r2b2/tests/util.py]: The hypothesis library is great for this sort of thing, and even tries to dig out corner cases in your code.
https://hypothesis.readthedocs.io/en/latest/

Perhaps we should some day move to using it, and put those tests in the r2b2/tests directory, and just hardcode a single testcase here.

File: src/r2b2/tests/election_template.json [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        "contests": {
+                "contest1": {
+                          "contest_ballots" : 100,
+                          "tally" : {
+                                  "CandidateA" : 50,
+                                  "CandidateB" : 50
+                          },
+                          "num_winners" : 1,
+                          "reported_winners" : ["CandidateA"],
+                          "contest_type" : "PLURALITY"
```
Comment (nealmcb) [inline src/r2b2/tests/election_template.json]: ValueError: 'PLURALITY' is not a valid ContestType
